degirum_tools/line_count.py

```diff
@@ -96,44 +121,97 @@ def analyze(self, result):
+            counts, trail_vector, line_vector, cross_product, trail_onto_line_projection
+        ):
+            if self._absolute_directions:
+                if trail_vector[0] < 0:
+                    counts.left += 1
+                else:
+                    counts.right += 1
+                if trail_vector[1] < 0:
+                    counts.top += 1
+                else:
+                    counts.bottom += 1
+            else:
+                if cross_product > 0:
+                    counts.left += 1
+                elif cross_product < 0:
+                    counts.right += 1
+                else:
+                    if np.sign(trail_vector) == np.sign(line_vector):
+                        counts.left += 1
+                    else:
+                        counts.right += 1
 
-        # obtain a set of new trails, which were not counted yet
-        new_trails = active_trails - self._counted_trails
+                trail_onto_line_projection_sign = np.sign(trail_onto_line_projection)
+                if np.all(trail_onto_line_projection_sign == np.sign(line_vector)):
+                    counts.top += 1
+                else:
+                    if not np.any(trail_onto_line_projection_sign):
+                        if cross_product > 0:
+                            counts.bottom += 1
+                        else:
+                            counts.top += 1
+                    else:
+                        counts.bottom += 1
 
-        def count_increment(counts, trail_start, trail_end):
-            if trail_start[0] > trail_end[0]:
-                counts.left += 1
-            else:
-                counts.right += 1
-            if trail_start[1] > trail_end[1]:
-                counts.top += 1
-            else:
-                counts.bottom += 1
+        if not self._accumulate:
+            self._line_counts = [LineCounts() for _ in self._lines]
 
         for tid in new_trails:
             trail = get_anchor_coordinates(
                 np.array(result.trails[tid]), self._anchor_point
             )
             if len(trail) > 1:
-                trail_start = trail[0]
+                trail_start = trail[0] if self._whole_trail else trail[-2]
                 trail_end = trail[-1]
 
-                for total_count, line in zip(self._line_counts, self._lines):
-                    if intersect(line[:2], line[2:], trail_start, trail_end):
-                        self._counted_trails.add(tid)
-                        count_increment(total_count, trail_start, trail_end)
+                for total_count, line, line_vector in zip(
+                    self._line_counts, self._lines, self._line_vectors
+                ):
+                    if intersect(line[0], line[1], trail_start, trail_end):
+                        if self._count_first_crossing:
+                            self._counted_trails.add(tid)
+                        trail_vector = self._line_to_vector((trail_start, trail_end))

```

Comment:
Makes sense to move trail_onto_line_projection, and cross_product calculations into count_increment(), because they are used only when not self._absolute_directions. Moving them down, will eliminate unnecessary calculations and will simplify count_increment() signature.